File: Interval Sum II.cpp
class SegmentTreeSumNode {
public:
    int start, end;
    long long sum;
    SegmentTreeSumNode *left, *right;
    SegmentTreeSumNode(int start, int end)
    {
        this->start = start;
        this->end = end;
        this->sum = sum;
        this->left = this->right = nullptr;
    }
};
class Solution {
// private:
    // SegmentTreeSumNode *root_;
public:
    /* you may need to use some attributes here */
    

    /**
     * @param A: An integer vector
     */
    Solution(vector<int> A):nums_(A) {
        // write your code here
        // root_ = build(0, A.size() - 1, A);
        bit_sum_ = vector<int>(nums_.size() + 1);
        for (int i = 0; i < nums_.size(); ++i)
        {
            add(i, nums_[i]);
        }
    }
    
    /**
     * @param start, end: Indices
     * @return: The sum from start to end
     */
    long long query(int start, int end) {
        // write your code here
        // return query(start, end, root_);
        int sum = sumRegin_bit(end);
        if (start > 0)
        {
            sum -= sumRegin_bit(start - 1);
        }
        return sum;
    }
    
    long long query(int start, int end, SegmentTreeSumNode *node)
    {
        if (!node || node->start > end || node->end < start)
        {
            return 0L;
        }
        else if (node->start >= start && node->end <= end)
        {
            return node->sum;
        }
        else
        {
            long long left = query(start, end, node->left);
            long long right = query(start, end, node->right);
            return left + right;
        }
    }
    /**
     * @param index, value: modify A[index] to value.
     */
    void modify(int index, int value) {
        // write your code here
        // modify(index, value, root_);
        add(index, value - nums_[index]);
        nums_[index] = value;
    }
    
    long long modify(int index, int value, SegmentTreeSumNode *node)
    {
        if (!node || node->start > index || node->end < index)
        {
            return node ? node->sum : 0;
        }
        else if (node->start == index && node->end == index)
        {
            node->sum = value;
            return value;
        }
        else
        {
            long long left = modify(index, value, node->left);
            long long right = modify(index, value, node->right);
            node->sum = left + right;
            return node->sum;
        }
    }
    
    SegmentTreeSumNode *build(int start, int end, vector<int> &A)
    {
        if (start > end)
        {
            return nullptr;
        }
        else
        {
            SegmentTreeSumNode *root = new SegmentTreeSumNode(start, end);
            if (start == end)
            {
                root->sum = A[start];
                return root;
            }
            else
            {
                int mid = start + (end - start) / 2;
                root->left = build(start, mid, A);
                root->right = build(mid + 1, end, A);
                long long left_sum = root->left ? root->left->sum : 0;
                long long right_sum = root->right ? root->right->sum : 0;
                root->sum = left_sum + right_sum;
                return root;
            }
        }
    }
    //solution II
    //hard to understand.
private:
    vector<int> nums_;
    vector<int> bit_sum_;
    int sumRegin_bit(int i)
    {
        ++i;
        int sum = 0;
        for(;i > 0; i -= lower_bit(i))
        {
            sum += bit_sum_[i];
        }
        return sum;
    }
    void add(int i, int value)
    {
        ++i;
        for(; i < bit_sum_.size(); i += lower_bit(i))
        {
            bit_sum_[i] += value;
        }
    }
    int lower_bit(int i)
    {
        return i & -i;
    }
};

File: Interval Sum II.py
class Solution:	
    class SegmentTree:
        def __init__(self, start, end, total):
            self.total = total
            self.start, self.end = start, end
            self.left, self.right = None, None
    def build(self, A):
        def doBuild(start, end, A):
            if start > end:
                return None
            if start == end:
                node = self.SegmentTree(start, end, A[start])
                return node
            node = self.SegmentTree(start, end, 0)
            mid = start + (end - start) / 2
            node.left = doBuild(start, mid, A)
            node.right = doBuild(mid + 1, end, A)
            node.total = node.left.total + node.right.total
            return node
            
        if not A:
            return None
        length = len(A)
        return doBuild(0, length - 1, A)
    # @param A: An integer list
    def __init__(self, A):
        # write your code here
        self.root = self.build(A)

    # @param start, end: Indices
    # @return: The sum from start to end
    def query(self, start, end):
        # write your code here
        def helper(node, start, end):
            if not node or node.start > end or node.end < start:
                return 0
            if node.start >= start and node.end <= end:
                return node.total
            return helper(node.left, start, end) + helper(node.right, start, end)
        return helper(self.root, start, end)
  
    # @param index, value: modify A[index] to value.
    def modify(self, index, value):
        # write your code here
        def helper(node, index, value):
            if not node:
                return 0
            if node.start > index or node.end < index:
                return node.total
            if node.start == index and node.end == index:
                node.total = value
                return value
            node.total = helper(node.left, index, value) + helper(node.right, index, value)
            return node.total
        return helper(self.root, index, value)
        